### PR TITLE
Fixed #28600 -- Added prefetch_related() support to RawQuerySet.

### PR DESCRIPTION
**Pull request by uadnan**
This PR adds

- Results caching support in `RawQuerySet`
- Prefetch related support in `RawQuerySet`

Ticket: https://code.djangoproject.com/ticket/28600

### CODE REVIEW

**Comment by uadnan** [comment]
@auvipy I've added whatever you have requested & also unset the flag on ticket. Can you please review?

File: docs/topics/db/sql.txt [hunk 1]
```diff
@@ -54,6 +54,23 @@ Of course, this example isn't very exciting -- it's exactly the same as
...
+Likewise :class:`~django.db.models.query.QuerySet`, ``RawQuerySet`` also
+implements internal results caching (see :ref:`caching-and-querysets`).
+
+``RawQuerySet`` supports specifying prefetch related lookups. If prefetch
+related lookup is defined on :class:`~django.db.models.query.QuerySet`
+that will be carry forward to ``RawQuerySet``.
+
+.. warning::
+
+    ``RawQuerySet`` internal caching is not backward incompatible. If you
```
Comment (timgraham) [inline docs/topics/db/sql.txt]: Notes about backwards incompatibilities  go in the release notes under "Backwards incompatible changes in 2.1".

File: docs/releases/2.1.txt [hunk 2]
```diff
@@ -155,6 +155,17 @@ Models
...
+* ``django.db.models.query.RawQuerySet`` now features internal
+  results caching (there exists no backward compatibility for this. If
+  hitting database each time ``RawQuerySet`` is executed, consider using
+  ``.iterator()`` on ``RawQuerySet``).
+* ``django.db.models.query.RawQuerySet`` now implements prefetch related
+  and carry forward prefetch_related lookups from
+  :class:`~django.db.models.query.QuerySet`.
+
+
+now implements ``__len__``,
```
Comment (timgraham) [inline docs/releases/2.1.txt]: There are some problems here.

File: docs/topics/db/sql.txt [hunk 3]
```diff
@@ -54,6 +54,23 @@ Of course, this example isn't very exciting -- it's exactly the same as
 running ``Person.objects.all()``. However, ``raw()`` has a bunch of other
 options that make it very powerful.
 
+
+.. versionadded:: 2.1
```
Comment (timgraham) [inline docs/topics/db/sql.txt]: The `.. versionadded:: 2.1` annotation should be self-contained as described in [Documenting new features](https://docs.djangoproject.com/en/dev/internals/contributing/writing-documentation/#documenting-new-features).

File: django/db/models/query.py [hunk 4]
```diff
@@ -1284,7 +1289,48 @@ def resolve_model_init_order(self):
...
+        prefetch_related_objects(self._result_cache, *self._prefetch_related_lookups)
+        self._prefetch_done = True
+
+    def _clone(self):
+        """
+        Return a copy of the current QuerySet. A lightweight alternative
+        to deepcopy().
+        """
+        c = self.__class__(self.raw_query, model=self.model, query=self.query, params=self.params,
+                           translations=self.translations, using=self._db, hints=self._hints)
```
Comment (timgraham) [inline django/db/models/query.py]: Use hanging indent as described in [Python coding style](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/coding-style/#python-style).

File: django/db/models/query.py [hunk 5]
```diff
@@ -1281,7 +1286,50 @@ def resolve_model_init_order(self):
...
 
+    def prefetch_related(self, *lookups):
+        """
+        Return a new RawQuerySet instance that will prefetch the specified
+        Many-To-One and Many-To-Many related objects when the RawQuerySet is
+        evaluated.
+
+        When prefetch_related() is called more than once, append to the list of
+        prefetch lookups. If prefetch_related(None) is called, clear the list.
+        """
```
Comment (carltongibson) [inline django/db/models/query.py]: Do we need to repeat the entire docstring from the QuerySet version?

File: docs/releases/2.1.txt [hunk 6]
```diff
@@ -370,6 +377,10 @@ Miscellaneous
 * ``django.contrib.staticfiles.templatetags.static()`` is deprecated in favor
   of ``django.templatetags.static.static()``.
 
+* ``RawQuerySet`` now implements internal results caching. If you want to hit
+  db each time ``RawQuerySet`` is iterated, consider using ``.iterator()``
```
Comment (carltongibson) [inline docs/releases/2.1.txt]: Not sure about "...hit db each time..."

File: docs/releases/2.1.txt [hunk 7]
```diff
@@ -221,6 +221,13 @@ Models
 * :meth:`.QuerySet.order_by` and :meth:`distinct(*fields) <.QuerySet.distinct>`
   now support using field transforms.
 
+* ``django.db.models.query.RawQuerySet`` now features internal
+  results caching.
+
+* ``django.db.models.query.RawQuerySet`` implements prefetch related
```
Comment (carltongibson) [inline docs/releases/2.1.txt]: `prefetch_related`.

File: docs/releases/2.1.txt [hunk 8]
```diff
@@ -221,6 +221,13 @@ Models
...
+  and carry forward prefetch_related lookups from
```
Comment (carltongibson) [inline docs/releases/2.1.txt]: Not sure about `carry forward`

**Comment by uadnan** [comment]
Hi @carltongibson I've added the requested unit test and also re-phrased the docs a little bit. Please have a look again when you have time

File: docs/releases/2.1.txt [hunk 9]
```diff
@@ -423,6 +428,9 @@ Miscellaneous
   accept ``obj`` as the second positional argument will be removed in Django
   3.0.
 
+* ``RawQuerySet`` now implements internal results caching, that can be avoided by
+  using ``.iterator()`` instead.
```
Comment (carltongibson) [inline docs/releases/2.1.txt]: This is in the wrong section. (It's in _Features deprecated... > Misc_ it need to be in _BC > Misc_, which is just above.)

I'd rephrase slightly. Something like: 

```
``django.db.models.query.RawQuerySet`` now implements internal results caching. You may use ``RawQuerySet.iterator()`` to bypass the results cache.
```

File: docs/topics/db/sql.txt (hunk 3 above)
Comment (carltongibson) [reply docs/topics/db/sql.txt]: This block is still not self contained, so that it can be removed after 2 versions.

File: docs/releases/2.1.txt [hunk 10]
```diff
@@ -235,6 +235,11 @@ Models
 * :class:`~django.db.models.BooleanField` can now be ``null=True``. This is
   encouraged instead of :class:`~django.db.models.NullBooleanField`, which will
   likely be deprecated in the future.
+* Internal results caching is enabled for ``django.db.models.query.RawQuerySet``.
```
Comment (carltongibson) [inline docs/releases/2.1.txt]: missing newline above this.

**Comment by uadnan** [comment]
@carltongibson `__len__` & `__bool__` are not implemented in this PR. I'll open a new PR for that and afterward and ll remove below warning in that PR.

    While a ``RawQuerySet`` instance can be iterated over like a normal
    :class:`~django.db.models.query.QuerySet`, ``RawQuerySet`` doesn't
    implement all methods you can use with ``QuerySet``. For example,
    ``__bool__()`` and ``__len__()`` are not defined in ``RawQuerySet``, and
    thus all ``RawQuerySet`` instances are considered ``True``. The reason
    these methods are not implemented in ``RawQuerySet`` is that implementing
    them without internal caching would be a performance drawback and adding
    such caching would be backward incompatible.

Sounds good or should I push those too in this PR?

**Comment by carltongibson** [comment]
@uadnan I think we need to adjust the warning twice. Once for this PR. Once for the other. (Imagine we have to revert one due to a regression.)

**Comment by uadnan** [comment]
@carltongibson got the point. I've removed the reason phrase from that warning.

**Comment by timgraham** [comment]
Could it be split into two commits? Result caching, then prefetch related?

**Comment by uadnan** [comment]
@timgraham done :)
